### PR TITLE
Fix for #91, #93.  Branched from upstream with rebase again

### PR DESCRIPTION
**Pull request by pablo-mayrgundter**
This fixes the bugs with uploaded models (#91, #93).

Last I heard from Antonio, IFCjs doesn't support unloading models and instead they're relying on reloading the page.  We're not doing this, but instead reloading the threejs scene and IfcViewerAPI when we upload a new model.  Somehow there is state being retained such that the new viewer instance thinks the new model has an ID of 1 when the API calls use the implicit ID, but 0 when it's passed explicitly.  To fix it, I just set the implicit (internal) ID to 0.

In the process of debugging, I refactored all call sites for viewer to prefer passing model instead where possible, to utilize the implicit ID semantics.  I also normalized arg org by always passing model before element.

Demo here, to try uploading model and prePicking:

  https://pablo-mayrgundter.github.io/Share/share/v/p/index.ifc

Also, this is the first CL I have initiated in VSCode with:

- Create branch from upstream
- Rebase from upstream

Maybe that's redundant.  But changelog looks great!  I'll try just Create branch from upstream next time.

### CODE REVIEW

File: src/Components/ItemPanel.jsx [hunk 1]
```diff
@@ -6,40 +6,45 @@ import MenuButton from '../Components/MenuButton'
 
 
 /**
- * @param {Object} viewer
+ * @param {Number} model
```
Comment (OlegMoshkovich) [inline src/Components/ItemPanel.jsx:9]: include the description for the argument?

File: src/Components/ItemPanel.jsx [hunk 2]
```diff
@@ -6,40 +6,45 @@ import MenuButton from '../Components/MenuButton'
...
+         open={showItemPanel}
+       />}
     </div>
   )
 }
 
 
 /**
- * @param {Object} viewer
+ * @param {Object} model
```
Comment (OlegMoshkovich) [inline src/Components/ItemPanel.jsx:34]: include the description for the arguments?

File: src/Components/ItemProperties.jsx [hunk 3]
```diff
@@ -10,24 +10,25 @@ import {stoi} from '../utils/strings'
 import Toggle from './Toggle'
 import ExpansionPanel from './ExpansionPanel'
 
+
 /**
  * ItemProperties displays IFC element properties and possibly PropertySets.
- * @param {Object} viewer
+ * @param {Object} model
```
Comment (OlegMoshkovich) [inline src/Components/ItemProperties.jsx:16]: include the description for the arguments?

File: src/Components/ItemProperties.jsx [hunk 4]
```diff
@@ -73,14 +74,14 @@ async function createPropertyTable(props, viewer, serial = 0, isPset = false) {
 
 
 /**
```
Comment (OlegMoshkovich) [inline src/Components/ItemProperties.jsx:76]: include descriptions?

File: src/Components/ItemProperties.jsx [hunk 5]
```diff
@@ -146,33 +147,33 @@ async function prettyProps(key, value, viewer, serial = 0) {
...
-          await deref(value, viewer, serial,
-              async (v, vwr, srl) => await createPropertyTable(v, vwr, srl)),
+          await deref(value, model, serial,
+              async (v, mdl, srl) => await createPropertyTable(mdl, v, srl)),
           serial)
   }
 }
 
 
 /**
```
Comment (OlegMoshkovich) [inline src/Components/ItemProperties.jsx:165]: include descriptions?

File: src/Components/ItemProperties.jsx [hunk 6]
```diff
@@ -207,13 +208,13 @@ async function quantities(key, quantitiesObj, viewer, serial) {
...
+ * @param {Object} model
  * @param {Array} eltArr
- * @param {Object} viewer
  * @param {Number} serial
  * @param {function} ifcToRowCb Callback to convert an IFC elt to a table row
  * @return {Object} The react component or null if fail
  */
-async function unpackHelper(eltArr, viewer, serial, ifcToRowCb) {
+async function unpackHelper(model, eltArr, serial, ifcToRowCb) {
   // HasProperties behaves a little special.
```
Comment (OlegMoshkovich) [inline src/Components/ItemProperties.jsx:218]: clearer comment?

File: src/Components/NavPanel.jsx [hunk 7]
```diff
@@ -10,7 +10,7 @@ import NodeOpen from '../assets/3D/nodeOpen.svg'
 
 
 /**
```
Comment (OlegMoshkovich) [inline src/Components/NavPanel.jsx:12]: include descriptions?

File: src/Containers/CadView.jsx [hunk 8]
```diff
@@ -62,38 +62,55 @@ export default function CadView({
...
   const [loadingMessage, setLoadingMessage] = useState()
+  const [model, setModel] = useState(null)
+
+  /* eslint-disable react-hooks/exhaustive-deps */
+  useEffect(() => {
+    onModelPath()
+  }, [modelPath])
+
+
+  useEffect(() => {
```
Comment (OlegMoshkovich) [inline src/Containers/CadView.jsx:73]: when does the viewer change? include a description of the use Effect?

File: src/Containers/CadView.jsx [hunk 9]
```diff
@@ -62,38 +62,55 @@ export default function CadView({
...
     setShowItemPanel(false)
     setViewer(initViewer(pathPrefix))
-    debug().log('CadView#useEffect[modelPath], done setting new viewer')
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [modelPath])
+    debug().log('CadView#onModelPath, done setting new viewer')
+  }
 
 
-  useEffect(() => {
```
Comment (OlegMoshkovich) [inline src/Containers/CadView.jsx:79]: This useEffect is really long.
just saying:)

File: src/Containers/CadView.jsx [hunk 10]
```diff
@@ -161,62 +186,49 @@ export default function CadView({
...
-    initSearch(rootElt, viewer)
+    initSearch(model, rootElt)
+    setRootElement(rootElt)
     setShowNavPanel(true)
   }
 
 
   /**
+   * @param {Object} m The IfcViewerAPI instance.
    * @param {Object} rootElt Root ifc elment for recursive indexing.
```
Comment (OlegMoshkovich) [inline src/Containers/CadView.jsx:208]: typo in the element.

File: src/Containers/CadView.jsx [hunk 11]
```diff
@@ -161,62 +186,49 @@ export default function CadView({
...
     setupLookupAndParentLinks(rootElt, elementsById)
     setDoubleClickListener()
-    initSearch(rootElt, viewer)
+    initSearch(model, rootElt)
+    setRootElement(rootElt)
     setShowNavPanel(true)
   }
 
 
   /**
```
Comment (OlegMoshkovich) [inline src/Containers/CadView.jsx:206]: include description?

File: src/utils/Ifc.js [hunk 12]
```diff
@@ -11,26 +11,27 @@ export function isTypeValue(obj) {
...
  * @param {Object} elt IFC element.
- * @param {Object} viewer Instance of a viewer.
  * @return {string} String representation of an IFC element type, e.g. 'IFCELEMENT'
  */
-export function getType(elt, viewer) {
-  const ifcMgr = viewer.IFC.loader.ifcManager
-  const id = elt.expressID
-  return ifcMgr.getIfcType(0, id)
+export function getType(model, elt) {
+  // return model.ifcManager.getIfcType(0, elt.expressID)
```
Comment (OlegMoshkovich) [inline src/utils/Ifc.js:22]: comment

File: src/utils/Ifc.test.js [hunk 13]
```diff
@@ -30,7 +30,8 @@ test('IfcManager.getType', () => {
     expressID: 1,
     Name: 'Building',
   }
-  expect(getType(elt, new MockViewer)).toEqual('IFCELEMENT')
+  // expect(getType(new MockViewer().IFC.loader.ifcManager, elt)).toEqual('IFCELEMENT')
```
Comment (OlegMoshkovich) [inline src/utils/Ifc.test.js:33]: comment

**Review by OlegMoshkovich — COMMENTED** [review]
wow - a lot of changes. 
left few comments

**Comment by pablo-mayrgundter** [comment]
I added a fix for #93.  Please take another look.

File: src/Components/ItemProperties.jsx (hunk 6 above)
Comment (OlegMoshkovich) [reply src/Components/ItemProperties.jsx:228]: It would help to include more description.

File: src/Components/ItemProperties.jsx [hunk 14]
```diff
@@ -245,59 +260,67 @@ async function unpackHelper(eltArr, viewer, serial, ifcToRowCb) {
...
- * @param {String} d2
- * @param {Number} serial
+ * @param {Object} d1 Table cell data 1
+ * @param {Object} d2 Table cell data 2
  * @return {Object} The react component
  */
-function Row({d1, d2, serial}) {
+function Row({d1, d2}) {
+  if (d1 === null || d1 === undefined ||
+      d1 === null || d1 === undefined) {
```
Comment (OlegMoshkovich) [inline src/Components/ItemProperties.jsx:290]: should be d2 - I think.

File: src/Containers/CadView.jsx [hunk 15]
```diff
@@ -161,62 +190,52 @@ export default function CadView({
...
-    setSelectedElements([])
-    if (viewer) {
-      viewer.IFC.unpickIfcItems()
-    }
-  }
-
-
   /**
    * Search for the query in the index and select matching items in UI elts.
    * @param {string} query The search query.
```
Comment (OlegMoshkovich) [inline src/Containers/CadView.jsx:230]: Query is obtained through parsing of the URL, not passed into onSearch function.Right?

**Review by OlegMoshkovich — COMMENTED** [review]
few comments.

**Comment by pablo-mayrgundter** [comment]
ptal

**Pull request merged**